apps/experiments/models.py

```diff
@@ -530,36 +530,31 @@ def get_data_for_experiment(self, experiment) -> dict:
             return {}
 
     @transaction.atomic()
-    def update_memory(self, data: dict, experiment: Experiment | None = None):
+    def update_memory(self, data: dict, experiment: Experiment):
         """
         Updates this participant's data records by merging `data` with the existing data. By default, data for all
-        experiments the this participant participated in will be updated. If there are no records for a specific
+        experiments that this participant participated in will be updated. If there are no records for a specific
         experiment, one will be created.
 
         Paramters
         data:
             A dictionary containing the new data
         experiment:
-            If specified, only the data for this experiment will be updated
+            Create a new record for this experiment if one does not exist
         """
-        experiments = Experiment.objects.filter(team=self.team).prefetch_related(
-            Prefetch("participant_data", queryset=ParticipantData.objects.filter(participant=self))
-        )
-        if experiment:
-            experiments = experiments.filter(id=experiment.id)
-
-        records_to_update = []
-        for experiment in experiments:
-            participant_data = experiment.participant_data.first()
-            # We cannot update the participant data using a single query, since the `data` field is encrypted at
-            # the application level
-            if participant_data:
-                participant_data.data = participant_data.data | data
-                records_to_update.append(participant_data)
-            else:
-                ParticipantData.objects.create(team=self.team, content_object=experiment, data=data, participant=self)
-
-        ParticipantData.objects.bulk_update(records_to_update, fields=["data"])
+        # Update all existing records
+        participant_data = ParticipantData.objects.filter(participant=self).select_for_update()
+        experiments = set()
+        with transaction.atomic():
+            experiment_content_type = ContentType.objects.get_for_model(Experiment)
```

Comment:
super nit (feel free to ignore, since this will not have a performance impact at this stage): we can update the pariticipant data query like this to avoid fetching unwanted results:

```python
experiment_content_type = ContentType.objects.get_for_model(Experiment)
ParticipantData.objects.filter(participant=self, content_type=experiment_content_type).select_for_update()
```